### ISSUE TITLE
Allow automatically rewriting any URL prefix

### ISSUE THREAD

**Issue by pimterry** [httptoolkit/httptoolkit#63]
HTTP Toolkit can currently automatically rewrite the host of request URLs, but not the entire URL, so this needs to be done manually with a breakpoint.

It would be useful if this worked automatically, so users could specify a URL  prefix (e.g. `example.com/api/v1/users`), and a replacement value (e.g. `http://localhost:8080/users`), so that requests to `https://example.com/api/v1/users/123` go to `https://localhost:8080/users/123`.

---

_Does this affect you too? Click below and add a :+1: to vote for this and help decide where HTTP Toolkit goes next, or [go vote on the other most popular ideas so far](https://github.com/httptoolkit/feedback/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc)._

**Comment by bpartridge** [comment]
This obviously wouldn't work for localhost, but for something on the public internet, you might want to look at redirecting the traffic to a Cloudflare worker which can edit arbitrary headers/paths and apply any logic you want!

**Comment by ri0ter** [comment]
@pimterry are you currently working on this feature or can I handle it?

**Comment by pimterry** [comment]
No, I'm not working on it, feel free! Bear in mind too that open-source contributors get [free HTTP Toolkit Pro](https://github.com/httptoolkit/httptoolkit#contributing-directly) :smiley:

Although I'm not working on this right now, I have had a look into it already, let me quickly add my notes so far:

* I think the way this should probably work is by extending the existing forwarding system, not adding a separate one (extending the "Forward the request to a different host" rule in the UI).
  * The internals of this are implemented in Mockttp [here](https://github.com/httptoolkit/mockttp/blob/2f5996627f1e9b67880d610c5c388a969cca3200/src/rules/handlers.ts#L664-L683), and tested [here](https://github.com/httptoolkit/mockttp/blob/2f5996627f1e9b67880d610c5c388a969cca3200/test/integration/proxy.spec.ts#L1377-L1455)
  * The UI to configure a forwarding rule is [here](https://github.com/httptoolkit/httptoolkit-ui/blob/abb45814fe9f0252cdfc8f39d496cce91e925983/src/components/mock/handler-config.tsx#L492-L615), and that then passes the configuration into Mockttp [here](https://github.com/httptoolkit/httptoolkit-ui/blob/abb45814fe9f0252cdfc8f39d496cce91e925983/src/model/rules/rule-definitions.ts#L132-L157)
* Right now that rule just forwards to a different host, with a single option to update/not update/override the host header.
* To allow more complex rewriting, the host should become optional, and you should be allowed to also add 0+ pairs of prefix paths + target paths that should be mapped. So something like "redirect to host example.com, mapping /v1/hello to /v2/hello and /v1/goodbye to /v2/adios".
  * I think the proxy implementation is fairly simple: find the first prefix that is a prefix of the request's real path, and replace that part of the path with the given target path.
  * On the UI side, you just need a way to easily enter the pairs there, which should be possible with a UI component very similar to the existing [editable headers](https://github.com/httptoolkit/httptoolkit-ui/blob/abb45814fe9f0252cdfc8f39d496cce91e925983/src/components/common/editable-headers.tsx) UI (used for entering header pairs for the fixed response rule, for example).

Give it a go, feel free to get in touch if you have any questions at all.

**Comment by pimterry** [comment]
This has now been fixed! In the latest UI release, within transform rules you can now use regular expressions to precisely match & replace any part of a request URL. Using regular expressions to match prefixes as described here and many other more complex or arbitrary transformations.

<img width="779" height="1049" alt="Image" src="https://github.com/user-attachments/assets/7660ff50-c9d5-44da-93bf-426107078b27" />
